### PR TITLE
net: add TcpListener.local_addr method

### PR DESCRIPTION
**Pull request by FrankReh**
The local_addr method is a blocking call.

### CODE REVIEW

**Comment by FrankReh** [comment]
Fixes TCP portion of #106 

But maybe there is a more standard way to cast the fd to a net::TcpListener?

File: src/net/tcp/listener.rs [hunk 1]
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
...
-    /// method.
     pub fn bind(addr: SocketAddr) -> io::Result<Self> {
         let socket = Socket::bind(addr, libc::SOCK_STREAM)?;
         socket.listen(1024)?;
         Ok(TcpListener { inner: socket })
     }
 
+    /// Returns the local address that this listener is bound to.
+    ///
+    /// This can be useful, for example, when binding to port 0 to figure out
```
Comment (Noah-Kennedy) [inline src/net/tcp/listener.rs:51]: I wouldn't exactly call this blocking. Certain system calls (like this one) are fast enough that they can be considered nonblocking. This one will simply drop to kernelspace to grab some data about the socket and return back.

It's probably even less expensive than another syscall we generally agree to be nonblocking, the `recv` syscall (in nonblocking mode), which does far more work in kernelspace than this one!

File: src/net/tcp/listener.rs [hunk 2]
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
     ///
     /// Binding with a port number of 0 will request that the OS assigns a port
     /// to this listener.
-    ///
```
Comment (Noah-Kennedy) [inline src/net/tcp/listener.rs:43]: Yeah I think this comment was kinda wrong tbh anyways.

File: src/net/tcp/listener.rs [hunk 3]
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
...
+    /// ```
+    /// use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
+    /// use tokio_uring::net::TcpListener;
+    ///
+    /// let listener = TcpListener::bind("127.0.0.1:8080".parse().unwrap()).unwrap();
+    ///
+    /// let addr = listener.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr, SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 8080)));
+    /// ```
+    #[cfg(unix)]
```
Comment (Noah-Kennedy) [inline src/net/tcp/listener.rs:66]: Not sure if the `#[cfg(unix)]` is necessary considering that this crate only works on unix in the first place.

**Review by Noah-Kennedy — CHANGES_REQUESTED** [review]

**Comment by Noah-Kennedy** [comment]
Some minor feedback, but overall this looks good!

File: src/net/tcp/listener.rs (hunk 1 above)
Comment (FrankReh) [reply src/net/tcp/listener.rs:51]: Agreed.

File: src/net/tcp/listener.rs (hunk 3 above)
Comment (FrankReh) [reply src/net/tcp/listener.rs:66]: I wasn't sure either but thought it safer to include so you could comment. Happy to remove it.

File: src/net/tcp/listener.rs [hunk 4]
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
...
+    pub fn local_addr(&self) -> io::Result<SocketAddr> {
+        use std::os::unix::io::{AsRawFd, FromRawFd};
+
+        let fd = self.inner.as_raw_fd();
+        // SAFETY: Our fd is the handle the kernel has given us for a TcpListener.
```
Comment (FrankReh) [inline src/net/tcp/listener.rs:71]: Is this SAFETY comment too much? Happy to replace if you want to suggest an alternative.

**Comment by FrankReh** [comment]
Think I handled your first set of comments.

I added a question about the SAFETY comment. Is it too wordy?

And importantly, I've added the local_addr method for the UnixListener now but have only tested it with the document example.

File: src/net/tcp/listener.rs (hunk 3 above)
Comment (Noah-Kennedy) [reply src/net/tcp/listener.rs]: Yeah I'd recommend removing it

File: src/net/tcp/listener.rs (hunk 4 above)
Comment (Noah-Kennedy) [reply src/net/tcp/listener.rs:69]: This is great!

**Review by Noah-Kennedy — APPROVED** [review]
